src/mcs/montecarlo.jl

```diff
@@ -75,106 +94,118 @@ function _store_param_results(m::Model, datum_key::Tuple{Symbol, Symbol}, trialn
+    savelist = sim_inst.sim_def.savelist
 
-    for (m, results) in zip(sim.models, sim.results)
+    model_index = 1
+    for (m, results) in zip(sim_inst.models, sim_inst.results)
         for datum_key in savelist
-            _store_param_results(m, datum_key, trialnum, results)
+            trial_df = _store_param_results(m, datum_key, trialnum, scen_name, results)
+            if output_dir !== nothing
+
+                # get sub_dir, which is different from output_dir if there are multiple models
+                if (length(sim_inst.results) > 1)
+                    sub_dir = joinpath(output_dir, "model_$(model_index)")
+                else
+                    sub_dir = output_dir   
+                end      
+                mkpath(sub_dir, mode=0o750) 
+
+                # get filtered trial_df, which is different from trial_df if there are multiple scenarios
+                if scen_name !== nothing
+                    trial_df_filtered = filter(row -> row[:scen] .== scen_name, trial_df)[:, 1:end-1] # remove scen field
+                else
+                    trial_df_filtered = trial_df
+                end
+
+                datum_name = join(map(string, datum_key), "_")
+                _save_trial_results(trial_df_filtered, datum_name, sub_dir, streams)
+            end
         end
+        model_index += 1
     end
 end
 
 """
-    save_trial_results(sim::Simulation, output_dir::String)
+    _save_trial_results(trial_df::DataFrame, datum_name::String, output_dir::String, streams::Dict{String, FileIO.Stream})
 
-Save the stored simulation results to files in the directory `output_dir`
+Save the stored simulation results in `trial_df` from trial `trialnum` to files in the directory `output_dir`
 """
-function save_trial_results(sim::Simulation{T}, output_dir::AbstractString) where T <: AbstractSimulationData
-    multiple_results = (length(sim.results) > 1)
-
-    mkpath(output_dir, mode=0o750)
-    
-    for (i, results) in enumerate(sim.results)
-        if multiple_results
-            sub_dir = joinpath(output_dir, "model_$i")
-            mkpath(sub_dir, mode=0o750)
-        else
-            sub_dir = output_dir 
-        end
-
-        for datum_key in sim.savelist
-            (comp_name, datum_name) = datum_key
-            filename = joinpath(sub_dir, "$datum_name.csv")
-            save(filename, results[datum_key])
-        end
+function _save_trial_results(trial_df::DataFrame, datum_name::String, output_dir::AbstractString, streams::Dict{String, FileIO.Stream}) where T <: AbstractSimulationData
+    filename = joinpath(output_dir, "$datum_name.csv")
+    if haskey(streams, filename)
+        CSVFiles.savestreaming(streams[filename], trial_df)
+    else
+        streams[filename] = CSVFiles.savestreaming(filename, trial_df)
     end
```

Comment:
This should work without the `CSVFiles.` before `savestreaming`. Without specifying `CSVFiles`, we can optionally save to any recognized file format that supports streaming.